### ISSUE TITLE
Upgrade to k8s 1.25 with PSS compliant applications

### ISSUE THREAD

**Issue by T-Kukawka** [giantswarm/roadmap#2716]

**Comment by T-Kukawka** [comment]
[Cluster-operator in next release](https://github.com/giantswarm/cluster-operator/pull/1643) will populate a flag {{- if (.global.podsecuritystandards.enforced) }} set to true that will have to be encapsulated in every PSP definition for every default application.
Default application meaning the apps tied to the release.

However this brings the problem with Managed Apps either deployed via Flux or App Platform. The mechanism we have right now with cluster-operator and cluster-values would work ONLY if customers do not add the ConfigMaps to actually 'personalize' the applications deployed. As we know customers add CMs to the applications we offer a lot. Adding ANYTHING in ConfigMap section will result in 'omitting' the cluster-values mounting in the App, hence omitting our flag that should be handled.

This creates several problems/different approaches which create more work in general:
- Either customers will have to edit all App CRs to mount the cluster-values config-map in the apps (to be done on MCs/Flux) so the switch is handled automatically based on the populated value from cluster-operator. This way the apps would 'migrate' during upgrade already to v20. However this creates A LOT of work for customers IMO.
- Or customers would have to e.g. change the flag in values.yaml(CM) per App such that the updated values are populated and handled while being on v20 release. This creates EVEN MORE work for customers and creates big room for missing the apps/settings etc.

What do we want to do? do we leave it in this state? Do we want to improve this first so cluster-values are added automatically to Managed Apps (if customized)?

Is current state acceptable? Can we move forward? (To be validated by @giantswarm/team-honeybadger )

**Comment by uvegla** [comment]
## Cluster values and purely relying on a global value

FTR Talking in the context of vintage clusters below (tho some parts would apply to CAPx too).

Apps can be and are deployed in many way. One major part of the need for manual checking from our side and the need for customer interaction too is that customer can deploy PSPs to the cluster too that they need to take care of. These PSPs do not necessarily come from GS Apps, they might be simply kubectl applied or just raw helm installed. This can be anything, we need to check in each case and ask customer to act / help them with it.

For Giant Swarm managed app it is multiple fold as well:

- there are GS managed apps that we deploy to the WC as part of the release. These manifest as App CRs in the cluster namespace
- we offer customer flux - flux-system namespace Flux, customers use default ns for resources usually, can differ. This is a vanilla flux, cannot by any means take advantage of `config` repo now as `konfigure` is not included in this flux. If the customer needs cluster values, they are responsible for mounting it for themselves
- install apps via Happa, which by default mounts the cluster values ([see](https://github.com/giantswarm/happa/blob/6593234400ea25b8afbdf336c10dc07e2d5434f1/src/components/MAPI/apps/utils.ts#L245))

Plus we have `app-admission-controller` on top of all that (all clusters run the latest release of it automatically) that has a mutating webhook that mounts the cluster values when the App CR's `.spec.config.configMap` entry is unpopulated ([see](https://github.com/giantswarm/app-admission-controller/blob/b5bcf7f4e33d7b26e930e8839d061ad7e83daa8d/pkg/app/mutate_app.go#L198-L200)). If some override is added there by the customer it is kept by the controller. Same for cluster secrets I think.

There is a special use case of the old `nginx` that until very recently had an exception in `app-admission-controller` to mount a different CM instead of the cluster values. ([see](https://github.com/giantswarm/app-admission-controller/releases/tag/v0.20.0) and [this](https://github.com/giantswarm/app-admission-controller/pull/321)) So this change is supposed to be released, but already existing apps will keep the old value, hook will see it has an override.

There are concerns around mounting the cluster values to everything (meaning really everything, whether it is customer or GS managed workload):

- you can potentially change / effect the release, deployment by overwriting values as the cluster values CM is not very well structured because the nodes are not unique / "namespaced" under a common let's say `giantswarm` node. Not saying it will definitely do, but can potentially affect workloads.
- there is no single or 2 or 3 ways to deploy things - cannot be guaranteed - so there cannot be a global / single place that can be 100% propagated everywhere
 - since we need a small set of value(s) now we could say we dont need to solve for the whole cluster values but just make that small thing available to everything to lower that side affect from the previous point, but the deployment issue applies here just the same

**Comment by piontec** [comment]
@T-Kukawka , @stone-z : is this summary from @uvegla good enough to move forward, as we discussed on Thursday?

**Comment by stone-z** [comment]
I'm interested in a potential alternative which wouldn't require as much effort. I'm hoping we can supply additional ConfigMaps (containing the PSP flag) to each App to be merged with any customer-configured values. [Thread](https://gigantic.slack.com/archives/C02GDJJ68Q1/p1692296040029919). But if that approach isn't feasible, then I'd need to re-enlist Honeybadger and/or Pheonix's help

**Comment by stone-z** [comment]
Lukasz and I met and discussed this today. I still want to follow up with Laszlo about the cluster values approach as I don't fully understand the implications yet and it seems there is some uncertainty involved. However, it seems that the feature Shield had in mind potentially meets another need Honeybadger has for app-operator, so it might be a win-win.
We have tentatively agreed:
1. Zach will sync with Laszlo to try to get my head around the cluster values option
2. If that solution is clearly and safely workable, we will go with that. If not:
3. Zach will propose a spec for the app-operator feature
4. Lukasz/Honeybadger will consider the spec and the necessary work
5. Either Honeybadger or Shield can implement it based on capacity and release timeline (v20 timing isn't certain yet)

**Comment by stone-z** [comment]
Laszlo and I met today and landed on a potentially even better solution. Shield will try using `app-admission-controller` to inject an additional `extraConfig` to each App CR, referencing a new ConfigMap in the WC namespace. This is exactly how cluster values are injected currently, is not reverted by Flux, and does not require updating any release components.

**Comment by T-Kukawka** [comment]
@stone-z how this will go with the plan to switch from PSP to PSS while upgrading to v20 (as done with cluster-operator)? Is this part of the app-operator then?

**Comment by stone-z** [comment]
> on which state of Values.global.podSecurityStandards.enforced should the PSP not be deployed?

When the value is `true`, PSPs must not be deployed. I agree the naming is confusing, I would have liked to make it a more explicit value like `global.podSecurityPolicy.enabled` but I think some teams have already implemented it so let's just keep it as-is.

There will be additional instructions in the issues I open for teams, cabbage is just ahead of the curve :broccoli: :1st_place_medal: 
We just want to actually test the theory before asking teams to implement anything and we're very overwhelmed with requests.

**Comment by mcharriere** [comment]
hi there. We've been testing this new value for cabbage apps and I noticed `cluster-operator v5.8` is already included in `aws v19.1`. 

That cluster-operator version already has the `global.podSecurityStandards.enforced` set to true; but the restricted PSP is still in place, consequently breaking any app that makes use of this value.

**Comment by T-Kukawka** [comment]
damn, we have updated components and missed that change, let me talk to the team

**Comment by T-Kukawka** [comment]
i guess this mostly affects the Managed Apps right @mcharriere ?

**Comment by mcharriere** [comment]
@T-Kukawka yes. default apps should not be affected (or we should have seen the problems already)

**Comment by T-Kukawka** [comment]
let me review the options

**Comment by T-Kukawka** [comment]
@mcharriere "Our best bet is to change that flag and make a patch release of cluster op, using the same release version" would that work for u?

**Comment by mcharriere** [comment]
> @mcharriere "Our best bet is to change that flag and make a patch release of cluster op, using the same release version" would that work for u?

that would be wonderful

**Comment by T-Kukawka** [comment]
fix has been released in the operator, it will soon be patched in the release as well

**Comment by T-Kukawka** [comment]
@othylmann the ticket description 'reflects' current understanding of the process we have in mind: https://github.com/giantswarm/roadmap/issues/2716#issue-1845195417
